src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+                            _threadHandling.ExecuteSynchronously(async () =>
+                            {
+                                foreach(var project in changedProjects)
+                                {
+                                    ProjectReloadResult result =  await project.ReloadProjectAsync().ConfigureAwait(true);
+
+                                    if(result == ProjectReloadResult.ReloadFailed || result == ProjectReloadResult.ReloadFailedProjectDirty)
+                                    {
+                                        failedProjects.Add(new Tuple<IReloadableProject, ProjectReloadResult>(project, result));
+                                    }
+                                }
+                            });
+
+                            ProcessProjectReloadFailures(failedProjects);
+
+                        });
+                    }
+                }
+            }
+            return VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// Prompts once for all the projects 
+        /// For each dirty project it prompts first to see what the user action needs to be and then
+        /// reloads the project if desired. Other options are do nothing amd save as and then reload
+        /// </summary>
+        private void ProcessProjectReloadFailures(List<Tuple<IReloadableProject, ProjectReloadResult>> failedProjects)
+        {
+            if(failedProjects.Count == 0)
+            {
+                return;
+            }
+
+            // Process each project. if the project is dirty in memory we need to prompt for each action. Non-dirty ones are
+            // simply reloaded
+            bool ignoreAll = false;
+            bool reloadAll = false;
+            foreach(var failure in failedProjects)
+            {
+                if(failure.Item2 == ProjectReloadResult.ReloadFailedProjectDirty)
+                {
+                    var buttons = new string[] {
+                                                Resources.Ignore,
+                                                Resources.Overwrite,
+                                                Resources.Discard,
+                                                Resources.SaveAs};
+                    var msgText = string.Format(Resources.ConflictingModificationsPrompt, Path.GetFileNameWithoutExtension(failure.Item1.ProjectFile));
+                    var dlg = new MultiChoiceMsgBox(Resources.ConflictingProjectModificationTitle, msgText, buttons);
+                    var dlgResult = dlg.ShowModal();
+                    if(dlgResult == true)
+                    {
+                        switch (dlg.SelectedAction)
+                        {
+                            case MultiChoiceMsgBoxResult.Cancel:
+                            {
+                                break;
+                            }
+                            case MultiChoiceMsgBoxResult.Button1:
```

Comment:
Nit: Could you throw a comment on each of these what case they're handling, so we don't have to look up and find the button list?